### PR TITLE
Typing the remaining scan requests

### PR DESCRIPTION
**Pull request by aryx**
This is typing the /scans/<scan_id/config request
and response and add in comments the type
of the response

test plan:
see related PR in semgrep

- [x] I ran `make setup && make` to update the generated code after editing a `.atd` file (TODO: have a CI check)
- [x] I made sure we're still backward compatible with old versions of the CLI.
      For example, the Semgrep backend need to still be able to *consume* data generated
	  by Semgrep 1.17.0.
      See https://atd.readthedocs.io/en/latest/atdgen-tutorial.html#smooth-protocol-upgrades

### CODE REVIEW

File: semgrep_output_v1.atd [hunk 1]
```diff
@@ -1008,6 +1056,12 @@ type ci_scan_results = {
    ?dependencies: ci_scan_dependencies option;
 }
 
+(* TODO: response to /results is {
+  errors: {message: string} list;
+  ?task_id: int option;
```
Comment (zyannes) [inline semgrep_output_v1.atd:1061]: This is a string type.

**Comment by aryx** [comment]
Diff from v1.17.0:
```
diff -u -p -b -B -r -x .semantic.cache -x .depend -x CVS -x .hg -x .svn -x .git -x _darcs /home/pad/github/FRESH-semgrep/cli/src/semgrep/app/scans.py /home/pad/yy/cli/src/semgrep/app/scans.py
--- /home/pad/github/FRESH-semgrep/cli/src/semgrep/app/scans.py	2023-10-03 10:29:31.157193643 +0200
+++ /home/pad/yy/cli/src/semgrep/app/scans.py	2023-10-03 09:27:29.146996852 +0200
@@ -3,27 +3,35 @@ import json
 import os
 from collections import Counter
 from copy import deepcopy
+from datetime import datetime
+from datetime import timedelta
 from pathlib import Path
+from time import sleep
 from typing import Any
 from typing import Dict
 from typing import FrozenSet
 from typing import List
 from typing import Optional
 from typing import Set
+from typing import Tuple
 from typing import TYPE_CHECKING
 from urllib.parse import urlencode
+from uuid import uuid4
 
 import click
 import requests
 from boltons.iterutils import partition
 
+import semgrep.semgrep_interfaces.semgrep_output_v1 as out
+from semdep.parsers.util import DependencyParserError
+from semgrep import __VERSION__
 from semgrep.constants import DEFAULT_SEMGREP_APP_CONFIG_URL
 from semgrep.constants import RuleSeverity
 from semgrep.error import SemgrepError
 from semgrep.parsing_data import ParsingData
+from semgrep.project import ProjectConfig
 from semgrep.rule import Rule
 from semgrep.rule_match import RuleMatchMap
-from semgrep.semgrep_interfaces.semgrep_output_v1 import FoundDependency
 from semgrep.state import get_state
 from semgrep.verbose_logging import getLogger
 
@@ -27,9 +35,9 @@ from semgrep.semgrep_interfaces.semgrep_
 from semgrep.state import get_state
 from semgrep.verbose_logging import getLogger
 
-
 if TYPE_CHECKING:
     from semgrep.engine import EngineType
+    from rich.progress import Progress
 
 logger = getLogger(__name__)
 
@@ -39,6 +47,12 @@ class ScanHandler:
         self._deployment_id: Optional[int] = None
         self._deployment_name: str = deployment_name
 
+        self.local_id = str(uuid4())
+        self.scan_metadata = out.ScanMetadata(
+            cli_version=out.Version(__VERSION__),
+            unique_id=self.local_id,
+            requested_products=[],
+        )
         self.scan_id = None
         self.ignore_patterns: List[str] = []
         self._policy_names: List[str] = []
@@ -50,6 +64,7 @@ class ScanHandler:
         self._skipped_match_based_ids: List[str] = []
         self._scan_params: str = ""
         self._rules: str = ""
+        self._enabled_products: List[str] = []
 
     @property
     def deployment_id(self) -> Optional[int]:
@@ -114,6 +129,13 @@ class ScanHandler:
         """
         return self._rules
 
+    @property
+    def enabled_products(self) -> List[str]:
+        """
+        Separate property for easy of mocking in test
+        """
+        return self._enabled_products
+
     def _get_scan_config_from_app(self, url: str) -> Dict[str, Any]:
         state = get_state()
         response = state.app_session.get(url)
@@ -152,10 +174,9 @@ class ScanHandler:
         if self.dry_run:
             app_get_config_url = f"{state.env.semgrep_url}/{DEFAULT_SEMGREP_APP_CONFIG_URL}?{self._scan_params}"
         else:
-            app_get_config_url = f"{state.env.semgrep_url}/{DEFAULT_SEMGREP_APP_CONFIG_URL}?{self._scan_params}"
-            # TODO: uncomment the line below to replace the old endpoint with the new one once we have the
-            # CLI logic in place to ignore findings that are from old rule versions
-            # app_get_config_url = f"{state.env.semgrep_url}/api/agent/deployments/scans/{self.scan_id}/config"
+            app_get_config_url = (
+                f"{state.env.semgrep_url}/api/agent/scans/{self.scan_id}/config"
+            )
 
         body = self._get_scan_config_from_app(app_get_config_url)
 
@@ -168,6 +189,7 @@ class ScanHandler:
         self._dependency_query = body.get("dependency_query") or False
         self._skipped_syntactic_ids = body.get("triage_ignored_syntactic_ids") or []
         self._skipped_match_based_ids = body.get("triage_ignored_match_based_ids") or []
+        self._enabled_products = body.get("enabled_products") or []
         self.ignore_patterns = body.get("ignored_files") or []
 
         if state.terminal.is_debug:
@@ -178,7 +200,9 @@ class ScanHandler:
                 pass
             logger.debug(f"Got configuration {json.dumps(config, indent=4)}")
 
-    def start_scan(self, meta: Dict[str, Any]) -> None:
+    def start_scan(
+        self, project_metadata: out.ProjectMetadata, project_config: ProjectConfig
+    ) -> None:
         """
         Get scan id and file ignores
 
@@ -189,10 +213,18 @@ class ScanHandler:
             logger.info(f"Would have sent POST request to create scan")
             return
 
-        logger.debug("Starting scan")
+        request = out.ScanRequest(
+            meta=out.RawJson(
+                {**project_metadata.to_json(), **project_config.to_dict()}
+            ),
+            scan_metadata=self.scan_metadata,
+            project_metadata=project_metadata,
+            project_config=out.ProjectConfig(out.RawJson(project_config.to_dict())),
+        ).to_json()
+        logger.debug(f"Starting scan: {json.dumps(request, indent=4)}")
         response = state.app_session.post(
             f"{state.env.semgrep_url}/api/agent/deployments/scans",
-            json={"meta": meta},
+            json=request,
         )
 
         if response.status_code == 404:
@@ -211,6 +243,7 @@ class ScanHandler:
 
         body = response.json()
         self.scan_id = body["scan"]["id"]
+        self._enabled_products = body["scan"].get("enabled_products") or []
 
     def report_failure(self, exit_code: int) -> None:
         """
@@ -246,15 +279,17 @@ class ScanHandler:
         parse_rate: ParsingData,
         total_time: float,
         commit_date: str,
-        lockfile_dependencies: Dict[str, List[FoundDependency]],
+        lockfile_dependencies: Dict[str, List[out.FoundDependency]],
+        dependency_parser_errors: List[DependencyParserError],
+        contributions: out.Contributions,
         engine_requested: "EngineType",
-    ) -> None:
+        progress_bar: "Progress",
+    ) -> Tuple[bool, str]:
         """
         commit_date here for legacy reasons. epoch time of latest commit
         """
         state = get_state()
-        all_ids = [r.id for r in rules]
-        cai_ids, rule_ids = partition(all_ids, lambda r_id: "r2c-internal-cai" in r_id)
+        rule_ids = [out.RuleId(r.id) for r in rules]
         all_matches = [
             match
             for matches_of_rule in matches_by_rule.values()
@@ -279,12 +314,8 @@ class ScanHandler:
         new_ignored, new_matches = partition(
             all_matches, lambda match: bool(match.is_ignored)
         )
-        findings = [
-            match.to_app_finding_format(commit_date).to_json() for match in new_matches
-        ]
-        ignores = [
-            match.to_app_finding_format(commit_date).to_json() for match in new_ignored
-        ]
+        findings = [match.to_app_finding_format(commit_date) for match in new_matches]
+        ignores = [match.to_app_finding_format(commit_date) for match in new_ignored]
         token = (
             # GitHub (cloud)
             os.getenv("GITHUB_TOKEN")
@@ -294,16 +325,20 @@ class ScanHandler:
             or os.getenv("BITBUCKET_TOKEN")
         )
 
-        findings_and_ignores = {
+        ci_scan_results = out.CiScanResults(
             # send a backup token in case the app is not available
-            "token": token,
-            "findings": findings,
-            "searched_paths": [str(t) for t in sorted(targets)],
-            "renamed_paths": [str(rt) for rt in sorted(renamed_targets)],
-            "rule_ids": rule_ids,
-            "cai_ids": cai_ids,
-            "ignores": ignores,
-        }
+            token=token,
+            findings=findings,
+            ignores=ignores,
+            searched_paths=[str(t) for t in sorted(targets)],
+            renamed_paths=[str(rt) for rt in sorted(renamed_targets)],
+            rule_ids=rule_ids,
+            contributions=contributions,
+        )
+        if self._dependency_query:
+            ci_scan_results.dependencies = out.CiScanDependencies(lockfile_dependencies)
+
+        findings_and_ignores = ci_scan_results.to_json()
 
         if any(match.severity == RuleSeverity.EXPERIMENT for match in new_ignored):
             logger.info("Some experimental rules were run during execution.")
@@ -315,71 +350,83 @@ class ScanHandler:
 
         dependency_counts = {k: len(v) for k, v in lockfile_dependencies.items()}
 
-        complete = {
-            "exit_code": 1
+        complete = out.CiScanComplete(
+            exit_code=1
             if any(match.is_blocking and not match.is_ignored for match in all_matches)
             else 0,
-            "stats": {
-                "findings": len(new_matches),
-                "errors": [error.to_dict() for error in errors],
-                "total_time": total_time,
-                "unsupported_exts": dict(ignored_ext_freqs),
-                "lockfile_scan_info": dependency_counts,
-                "parse_rate": {
-                    lang: {
-                        "targets_parsed": data.num_targets - data.targets_with_errors,
-                        "num_targets": data.num_targets,
-                        "bytes_parsed": data.num_bytes - data.error_bytes,
-                        "num_bytes": data.num_bytes,
-                    }
+            dependency_parser_errors=dependency_parser_errors,
+            stats=out.CiScanCompleteStats(
+                findings=len(
+                    [match for match in new_matches if not match.from_transient_scan]
+                ),
+                errors=[error.to_CliError() for error in errors],
+                total_time=total_time,
+                unsupported_exts=dict(ignored_ext_freqs),
+                lockfile_scan_info=dependency_counts,
+                parse_rate={
+                    lang: out.ParsingStats(
+                        targets_parsed=data.num_targets - data.targets_with_errors,
+                        num_targets=data.num_targets,
+                        bytes_parsed=data.num_bytes - data.error_bytes,
+                        num_bytes=data.num_bytes,
+                    )
                     for (lang, data) in parse_rate.get_errors_by_lang().items()
                 },
-                "engine_requested": engine_requested.name,
-            },
-        }
-
-        if self._dependency_query:
-            lockfile_dependencies_json = {}
-            for path, dependencies in lockfile_dependencies.items():
-                lockfile_dependencies_json[path] = [
-                    dependency.to_json() for dependency in dependencies
-                ]
-            complete["dependencies"] = lockfile_dependencies_json
+                engine_requested=engine_requested.name,
+            ),
+        )
 
         if self.dry_run:
             logger.info(
                 f"Would have sent findings and ignores blob: {json.dumps(findings_and_ignores, indent=4)}"
             )
             logger.info(
-                f"Would have sent complete blob: {json.dumps(complete, indent=4)}"
+                f"Would have sent complete blob: {json.dumps(complete.to_json(), indent=4)}"
             )
-            return
+            return (False, "")
         else:
             logger.debug(
                 f"Sending findings and ignores blob: {json.dumps(findings_and_ignores, indent=4)}"
             )
-            logger.debug(f"Sending complete blob: {json.dumps(complete, indent=4)}")
+            logger.debug(
+                f"Sending complete blob: {json.dumps(complete.to_json(), indent=4)}"
+            )
 
+        results_task = progress_bar.add_task("Uploading scan results")
         response = state.app_session.post(
-            f"{state.env.semgrep_url}/api/agent/scans/{self.scan_id}/findings_and_ignores",
+            f"{state.env.semgrep_url}/api/agent/scans/{self.scan_id}/results",
+            timeout=state.env.upload_findings_timeout,
             json=findings_and_ignores,
         )
 
         try:
             response.raise_for_status()
 
-            resp_errors = response.json()["errors"]
+            res = response.json()
+            resp_errors = res["errors"]
             for error in resp_errors:
                 message = error["message"]
                 click.echo(f"Server returned following warning: {message}", err=True)
 
-        except requests.RequestException:
-            raise Exception(f"API server returned this error: {response.text}")
+            if "task_id" in res:
+                complete.task_id = res["task_id"]
+
+            progress_bar.update(results_task, completed=100)
+
+        except requests.RequestException as exc:
+            raise Exception(f"API server returned this error: {response.text}") from exc
+
+        try_until = datetime.now() + timedelta(minutes=20)
+        slow_down_after = datetime.now() + timedelta(minutes=2)
+        complete_task = progress_bar.add_task("Finalizing scan")
+        while datetime.now() < try_until:
+            logger.debug("Sending /complete")
 
         # mark as complete
         response = state.app_session.post(
             f"{state.env.semgrep_url}/api/agent/scans/{self.scan_id}/complete",
-            json=complete,
+                timeout=state.env.upload_findings_timeout,
+                json=complete.to_json(),
         )
 
         try:
@@ -388,3 +435,20 @@ class ScanHandler:
             raise Exception(
                 f"API server at {state.env.semgrep_url} returned this error: {response.text}"
             )
+
+            ret = response.json()
+
+            if ret.get("success", False):
+                progress_bar.update(complete_task, completed=100)
+                return (
+                    ret.get("app_block_override", False),
+                    ret.get("app_block_reason", ""),
+                )
+            else:
+                progress_bar.advance(complete_task)
+                sleep(5 if datetime.utcnow() < slow_down_after else 30)
+                continue
+
+        raise Exception(
+            f"API server at {state.env.semgrep_url} never completed processing the scan results."
+        )

Diff finished.  Tue Oct  3 10:30:01 2023
```

File: semgrep_output_v1.atd [hunk 2]
```diff
@@ -919,6 +929,55 @@ type scan_request = {
...
+}
+type scan_response_bis = {
+    (* the scan id *)
+    id: string;
+    (* since 1.37.0 *)
+    ?enabled_products: product list;
+}
+*)
+
+(* TODO? seems used only for dry_run and passed urlencoded to /scans/config?xxx
```
Comment (jbergler) [inline semgrep_output_v1.atd:945]: Note to dig into this: I thought this re-used the same data as the scan create payload

File: semgrep_output_v1.atd [hunk 3]
```diff
@@ -919,6 +929,55 @@ type scan_request = {
...
+
+(* Response by the backend to the CLI to the POST /scans/<scan_id>/config *)
+type scan_config = {
+    deployment_id: int;
+    (* the org name *)
+    deployment_name: string;
+    (* ex: "audit", "comment", "block" TODO use enum? *)
+    policy_names: string list;
+    (* rules raw content in JSON format *)
+    rule_config: string;
```
Comment (jbergler) [inline semgrep_output_v1.atd:965]: This isn't always a string - I started migrating this to json for semgrep scan a while back and put the brakes on after that. Will dig into this today.

**Review by IagoAbal — APPROVED** [review]
stamp

File: semgrep_output_v1.atd (hunk 2 above)
Comment (aryx) [reply semgrep_output_v1.atd:945]: Yes it's weird we have such a complex structure that we just use for a dry_run ...

File: semgrep_output_v1.atd (hunk 3 above)
Comment (aryx) [reply semgrep_output_v1.atd:965]: We should not use different types ... In any case we could add another field with rule_as_json:

File: semgrep_output_v1.atd (hunk 1 above)
Comment (aryx) [reply semgrep_output_v1.atd]: done

**Pull request merged**